dbms/src/Storages/StorageDeltaMerge.cpp

```diff
@@ -1198,59 +1198,26 @@ void StorageDeltaMerge::rename(
     // For DatabaseTiFlash, simply update store's database is OK.
     // `store->getTableName() == new_table_name` only keep for mock test.
     bool clean_rename = !data_path_contains_database_name && getTableName() == new_table_name;
-    if (likely(clean_rename))
+    RUNTIME_ASSERT(clean_rename,
+                   log,
+                   "should never rename the directories when renaming table, new_database_name={}, new_table_name={}",
+                   new_database_name,
+                   new_table_name);
+    if (storeInited())
     {
-        if (storeInited())
-        {
-            _store->rename(new_path_to_db, clean_rename, new_database_name, new_table_name);
-            return;
-        }
-        std::lock_guard lock(store_mutex);
-        if (storeInited())
-        {
-            _store->rename(new_path_to_db, clean_rename, new_database_name, new_table_name);
-        }
-        else
-        {
-            table_column_info->db_name = new_database_name;
-            table_column_info->table_name = new_table_name;
-        }
+        _store->rename(new_path_to_db, clean_rename, new_database_name, new_table_name);
         return;
     }
-
-    /// Note that this routine is only left for CI tests. `clean_rename` should always be true in production env.
-    auto & store = getAndMaybeInitStore();
-
-    // For DatabaseOrdinary, we need to rename data path, then recreate a new store.
-    const String new_path = new_path_to_db + "/" + new_table_name;
-
-    if (Poco::File{new_path}.exists())
-        throw Exception(
-            fmt::format("Target path already exists: {}", new_path),
-            ErrorCodes::DIRECTORY_ALREADY_EXISTS);
-
-    // flush store and then reset store to new path
-    store->flushCache(global_context, RowKeyRange::newAll(is_common_handle, rowkey_column_size));
-    ColumnDefines table_column_defines = store->getTableColumns();
-    ColumnDefine handle_column_define = store->getHandle();
-    DeltaMergeStore::Settings settings = store->getSettings();
-
-    // remove background tasks
-    store->shutdown();
-    // rename directories for multi disks
-    store->rename(new_path, clean_rename, new_database_name, new_table_name);
-    // generate a new store
-    store = std::make_shared<DeltaMergeStore>(
-        global_context,
-        data_path_contains_database_name,
-        new_database_name,
-        new_table_name,
-        tidb_table_info.id,
-        std::move(table_column_defines),
-        std::move(handle_column_define),
-        is_common_handle,
-        rowkey_column_size,
-        settings);
```

Comment:
Use RUNTIME_ASSERT to disable this logic. Cause we never expect running to this branch.
And after this PR, it will remove the callbacks, also cleaning the external ids ptrs from PageStorage. After restoring the DeltaMergeStore from disk, it will run into remove all dtfiles from disk.

Reply:
Likely, we should not call shutdown for a table but later restore the DeltaMergeStoer instance

Reply:
The deleted cases are rewrite as unit test in https://github.com/pingcap/tiflash/pull/5762